File: Algorithm/java/include/globalFunc.java
package include;
public class globalFunc {
    // 产生一个随机数组，数组长度为n，数组元素的最大值为V
    public static int[] randomArray(int n, int V){
        int[] arr = new int[n];
        for(int i = 0; i < n; i++){
            arr[i] = (int)(Math.random() * V + 1);
        }

        return arr;
    }

    public static void printArray(int[] arr){
        System.out.print("[");
        for(int i = 0; i < arr.length - 1; i++){
            System.out.print(arr[i] + ", ");
        }
        System.out.print(arr[arr.length - 1] + "]\n");
    }

    // 复制数组，传入一个数组，复制该数组并返回
    public static int[] copyArray(int[] arr){

        int n = arr.length;
        int[] newarr = new int[n];
        for(int i = 0; i < n; i++){
            newarr[i] = arr[i];
        }

        return newarr;
    }

    // 返回两个数组是否相等的布尔值
    public static boolean sameArray(int[] arr1, int[] arr2){

        int n = arr1.length;
        for(int i = 0; i < n; i++){
            if(arr1[i] != arr2[i]){
                return false;
            }
        }

        return true;
    }

    // 交换数组中i和j位置的元素
    public static void swap(int[] arr, int i, int j){

        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp; 
    }


    public static class Node {
    
        public int value;
        public Node next;
        public Node rand;
        public Node(int data){
            this.value = data;
        }
    }

    // 初始化链表
    public static void initialNode(Node head){
        head.next = null;
    }

    // 使用尾插法插入一个元素到链表中
    public static void tailInsert(Node head, int e){
        Node node = new Node(e);
        Node p = head;
        while (p.next != null) {
            p = p.next;
        }
        p.next = node;
        node.next = null;
    }

    // 使用头插法插入一个元素到链表中
    public static void headInsert(Node head, int e){
        if(head == null){
            System.out.println("出现错误，请检查头节点");
        }
        Node node = new Node(e);
        if(head.next == null){
            head.next = node;
        }
        else{
            node.next = head.next;
            head.next = node;
        }

    }

    public static void printLinkList(Node head){
        if(head == null){
            System.out.println("链表为空！");
        }
        Node p = head;
        while (p != null) {
            System.out.print(p.value + " ");
            p = p.next;
        }
        System.out.println();
    }

    // 产生一个随机链表，链表长度为n，链表的结点的最大值为不超过V
    public static Node createRandomLinkList(int n, int V){

        Node head = new Node((int)(Math.random() * V) + 1);
        for(int i = 1; i < n; i++){
            headInsert(head, (int)(Math.random() * V) + 1);
        }

        return head;
    }

}

File: Algorithm/python/class02/globalFunc.py
from random import random

# 产生一个随机数组，数组的长度为n，数组元素的最大值为V
def randomArray(n, V):
    return [(int)(random() * V + 1) for _ in range(n)]


def copyArray(arr):
    return [x for x in arr]

def sameArray(arr1, arr2):
    return True if arr1 == arr2 else False

def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]

def selectSort(arr):
    for i in range(len(arr) - 1):
        minIndex = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[minIndex]:
                minIndex = j
        swap(arr, i, minIndex)


# 冒泡排序
def bubbleSort(arr):
    for end in range(len(arr), 1, -1):
        for j in range(0, end - 1):
            if arr[j] > arr[j + 1]:
                swap(arr, j, j + 1)

# 插入排序
def insertSort(arr):
    for i in range(1, len(arr)):
        for j in range(i, 0, -1):
            if arr[j] < arr[j - 1]:
                swap(arr, j, j - 1)

def mergeSort(arr):
    if arr == None or len(arr) < 2:
        return
    process(arr, 0, len(arr) - 1)


def process(arr, L, R):
    if L == R:      # base case
        return
    mid = L + ((R - L) >> 1)
    process(arr, L, mid)
    process(arr, mid + 1, R)
    merge(arr, L, mid, R)

def merge(arr, L, M, R):
    help = [0 for _ in range(R - L + 1)]
    i = 0
    p1 = L
    p2 = M + 1
    while p1 <= M and p2 <= R:
        if arr[p1] <= arr[p2]:
            help[i] = arr[p1]
            i, p1 = i + 1, p1 + 1
        else:
            help[i] = arr[p2]
            i, p2 = i + 1, p2 + 1

    while p1 <= M:
        help[i] = arr[p1]
        i, p1 = i + 1, p1 + 1

    while p2 <= R:
        help[i] = arr[p2]
        i, p2 = i + 1, p2 + 1
    
    for i in range(len(help)):
        arr[L + i] = help[i]
